### PR TITLE
chore: basic object tracking part 2

### PR DESCRIPTION
**Pull request by sfc-gh-jcieslak**
follow up to https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/3205
## Changes
- applied comments from the previous pr (added test in resource and account locator key was changed to `struct{}`, etc.)
- applied tracking wrappers to every resource that has implemented context method (in old ones we create context ourselves; I can go and add the context there too), also imports implemented as passthrough were skipped.
- added missing resource names to `resource` type
- added entry to our v1 guidelines to add wrapper around terraform functions
- modified test client for information schema a bit

## Coming up
- data sources (after pr that adds data source names similar to the `resrouces.Resource` interface)

### CODE REVIEW

File: pkg/acceptance/helpers/information_schema_client.go [hunk 1]
```diff
@@ -25,20 +27,37 @@ func (c *InformationSchemaClient) client() *sdk.Client {
...
+		return QueryHistory{
+			QueryId:   (*m["QUERY_ID"]).(string),
+			QueryText: (*m["QUERY_TEXT"]).(string),
+			QueryTag:  (*m["QUERY_TAG"]).(string),
+		}
+	})
 }
 
-func (c *InformationSchemaClient) GetQueryTagByQueryId(t *testing.T, queryId string) string {
+func (c *InformationSchemaClient) GetQueryHistoryByQueryId(t *testing.T, limit int, queryId string) QueryHistory {
```
Comment (sfc-gh-jmichalak) [inline pkg/acceptance/helpers/information_schema_client.go:52]: I think we can agree on a default limit here and avoid specifying it in each call.

File: pkg/resources/function.go [hunk 2]
```diff
@@ -166,15 +168,15 @@ func Function() *schema.Resource {
...
 
-		CustomizeDiff: customdiff.All(
+		CustomizeDiff: TrackingCustomDiffWrapper(resources.Function, customdiff.All(
 			// TODO(SNOW-1348103): add `arguments` to ComputedIfAnyAttributeChanged. This can't be done now because this function compares values without diff suppress.
 			ComputedIfAnyAttributeChanged(functionSchema, FullyQualifiedNameAttributeName, "name"),
-		),
+		)),
 
 		Schema: functionSchema,
 		Importer: &schema.ResourceImporter{
```
Comment (sfc-gh-jmichalak) [inline pkg/resources/function.go:182]: What about this import?

File: pkg/resources/network_rule.go [hunk 3]
```diff
@@ -65,10 +67,10 @@ var networkRuleSchema = map[string]*schema.Schema{
...
-		ReadContext:   ReadContextNetworkRule,
-		UpdateContext: UpdateContextNetworkRule,
-		DeleteContext: DeleteContextNetworkRule,
+		CreateContext: TrackingCreateWrapper(resources.NetworkRule, CreateContextNetworkRule),
+		ReadContext:   TrackingReadWrapper(resources.NetworkRule, ReadContextNetworkRule),
+		UpdateContext: TrackingUpdateWrapper(resources.NetworkRule, UpdateContextNetworkRule),
+		DeleteContext: TrackingDeleteWrapper(resources.NetworkRule, DeleteContextNetworkRule),
 
 		Schema: networkRuleSchema,
 		Importer: &schema.ResourceImporter{
```
Comment (sfc-gh-jmichalak) [inline pkg/resources/network_rule.go:76]: What about this import?

File: pkg/resources/cortex_search_service.go [hunk 4]
```diff
@@ -81,10 +83,10 @@ var cortexSearchServiceSchema = map[string]*schema.Schema{
...
-		ReadContext:   ReadCortexSearchService,
-		UpdateContext: UpdateCortexSearchService,
-		DeleteContext: DeleteCortexSearchService,
+		CreateContext: TrackingCreateWrapper(resources.CortexSearchService, CreateCortexSearchService),
+		ReadContext:   TrackingReadWrapper(resources.CortexSearchService, ReadCortexSearchService),
+		UpdateContext: TrackingUpdateWrapper(resources.CortexSearchService, UpdateCortexSearchService),
+		DeleteContext: TrackingDeleteWrapper(resources.CortexSearchService, DeleteCortexSearchService),
 
 		Schema: cortexSearchServiceSchema,
 		Importer: &schema.ResourceImporter{
```
Comment (sfc-gh-jmichalak) [inline pkg/resources/cortex_search_service.go:92]: What about this import?

File: pkg/resources/stage.go [hunk 5]
```diff
@@ -100,10 +102,10 @@ var stageSchema = map[string]*schema.Schema{
...
-		ReadContext:   ReadStage,
-		UpdateContext: UpdateStage,
-		DeleteContext: DeleteStage,
+		CreateContext: TrackingCreateWrapper(resources.Stage, CreateStage),
+		ReadContext:   TrackingReadWrapper(resources.Stage, ReadStage),
+		UpdateContext: TrackingUpdateWrapper(resources.Stage, UpdateStage),
+		DeleteContext: TrackingDeleteWrapper(resources.Stage, DeleteStage),
 
 		Schema: stageSchema,
 		Importer: &schema.ResourceImporter{
```
Comment (sfc-gh-jmichalak) [inline pkg/resources/stage.go:111]: What about this import?

File: pkg/resources/table.go [hunk 6]
```diff
@@ -211,9 +213,9 @@ func Table() *schema.Resource {
...
 		Delete: DeleteTable,
 
-		CustomizeDiff: customdiff.All(
+		CustomizeDiff: TrackingCustomDiffWrapper(resources.Table, customdiff.All(
 			ComputedIfAnyAttributeChanged(tableSchema, FullyQualifiedNameAttributeName, "name"),
-		),
+		)),
 
 		Schema: tableSchema,
 		Importer: &schema.ResourceImporter{
```
Comment (sfc-gh-jmichalak) [inline pkg/resources/table.go:221]: What about this import?

File: pkg/acceptance/helpers/information_schema_client.go [hunk 7]
```diff
@@ -25,20 +27,37 @@ func (c *InformationSchemaClient) client() *sdk.Client {
...
 	t.Helper()
-	result, err := c.client().QueryUnsafe(context.Background(), fmt.Sprintf("SELECT QUERY_TAG FROM TABLE(INFORMATION_SCHEMA.QUERY_HISTORY(RESULT_LIMIT => 20)) WHERE QUERY_ID = '%s'", queryId))
+	result, err := c.client().QueryUnsafe(context.Background(), fmt.Sprintf("SELECT QUERY_TEXT FROM TABLE(INFORMATION_SCHEMA.QUERY_HISTORY(RESULT_LIMIT => %d)) WHERE QUERY_ID = '%s'", limit, queryId))
 	require.NoError(t, err)
 	require.Len(t, result, 1)
-	require.NotNil(t, result[0]["QUERY_TAG"])
-	return (*result[0]["QUERY_TAG"]).(string)
+	require.NotNil(t, result[0]["QUERY_TEXT"])
+	return QueryHistory{
+		QueryId:   (*result[0]["QUERY_ID"]).(string),
```
Comment (sfc-gh-jmichalak) [inline pkg/acceptance/helpers/information_schema_client.go:59]: 1. Here we only select QUERY_TEXT, so why we set QueryTag here?
2. Why do we use a limit 20 and expect 1 result?
3. If we want to use the other columns, we should add `require.NotNil`, as above.

File: pkg/resources/usage_tracking_acceptance_test.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+	}) // Delete
+
+	assertQueryMetadataExists := func(t *testing.T, queryHistory []helpers.QueryHistory, operation tracking.Operation, query string) {
+		t.Helper()
+		found := false
+		expectedMetadata := tracking.NewVersionedMetadata(resources.User, operation)
+		for _, history := range queryHistory {
+			if metadata, err := tracking.ParseMetadata(history.QueryText); err == nil {
+				if expectedMetadata == metadata && strings.Contains(history.QueryText, query) {
+					found = true
```
[the remark this answers is not in the document]
Comment (sfc-gh-jmichalak) [reply pkg/resources/usage_tracking_acceptance_test.go:76]: I think we can use `collections.FindFirst` here.

**Review by sfc-gh-jmichalak — CHANGES_REQUESTED** [review]
Please also check other resources if they have import tracking set up.

File: pkg/provider/resources/resources.go [hunk 9]
```diff
@@ -23,6 +23,13 @@ const (
 	ExternalVolume                                         resource = "snowflake_external_volume"
 	FailoverGroup                                          resource = "snowflake_failover_group"
 	FileFormat                                             resource = "snowflake_file_format"
+	GrantApplicationRole                                   resource = "snowflake_grant_application_role"
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/provider/resources/resources.go:26]: Moved

File: pkg/sdk/client.go [hunk 10]
```diff
@@ -266,9 +266,9 @@ func (c *Client) Close() error {
 	return nil
 }
 
-type ContextKey string
+type ContextKey struct{}
 
-const SnowflakeAccountLocatorContextKey ContextKey = "snowflake_account_locator"
+var SnowflakeAccountLocatorContextKey ContextKey
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/sdk/client.go:271]: No, unexported and changed the type name to specify it is for the account locator

File: pkg/resources/usage_tracking_acceptance_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/resources/usage_tracking_acceptance_test.go:76]: Done

File: pkg/resources/usage_tracking_acceptance_test.go [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+	"github.com/hashicorp/terraform-plugin-testing/tfversion"
+
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+)
+
+func TestAcc_CompleteUsageTracking(t *testing.T) {
+	id := acc.TestClient().Ids.RandomAccountObjectIdentifier()
+	comment := random.Comment()
+
+	userModel := model.User("test", id.Name())
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/resources/usage_tracking_acceptance_test.go:27]: I wanted to use something with parameters, database and schema didn't have their assertions generated so I went with user. Changed to database.

File: pkg/resources/usage_tracking_acceptance_test.go [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+					found = true
+				}
+			}
+		}
+		if !found {
+			t.Fatalf("query history does not contain query metadata: %v with query containing: %s", expectedMetadata, query)
+		}
+	}
+
+	queryHistory := acc.TestClient().InformationSchema.GetQueryHistory(t, 100)
```
[the remark this answers is not in the document]
Comment (sfc-gh-jcieslak) [reply pkg/resources/usage_tracking_acceptance_test.go:85]: Done

File: pkg/acceptance/helpers/information_schema_client.go (hunk 1 above)
Comment (sfc-gh-jcieslak) [reply pkg/acceptance/helpers/information_schema_client.go:52]: I would leave it as a value to specify as I changed this a lot during testing and we should aim for as little number as possible

File: pkg/resources/function.go (hunk 2 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/function.go:182]: wdym? Import function is wrapped

File: pkg/resources/network_rule.go (hunk 3 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/network_rule.go:76]: passthrough imports are not wrapped

File: pkg/resources/cortex_search_service.go (hunk 4 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/cortex_search_service.go:92]: passthrough imports are not wrapped

File: pkg/resources/stage.go (hunk 5 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/stage.go:111]: passthrough imports are not wrapped

File: pkg/acceptance/helpers/information_schema_client.go (hunk 7 above)
Comment (sfc-gh-jcieslak) [reply pkg/acceptance/helpers/information_schema_client.go:59]: Added common mapper that has `require.NotNil`, limit is on the number of queries we look through, so to find this one query by query id we have look through `limit` number of queries in query history

File: pkg/resources/usage_tracking_acceptance_test.go [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+		return func(state *terraform.State) error {
+			queryHistory := acc.TestClient().InformationSchema.GetQueryHistory(t, 30)
+			expectedMetadata := tracking.NewVersionedMetadata(resources.Schema, operation)
+			if _, err := collections.FindFirst(queryHistory, func(history helpers.QueryHistory) bool {
+				if metadata, err := tracking.ParseMetadata(history.QueryText); err == nil {
+					if expectedMetadata == metadata && strings.Contains(history.QueryText, query) {
+						return true
+					}
+				}
+				return false
```
Comment (sfc-gh-jmichalak) [inline pkg/resources/usage_tracking_acceptance_test.go:48]: Nit:
```suggestion
				metadata, err := tracking.ParseMetadata(history.QueryText)
				return err == nil &&
				        expectedMetadata == metadata &&
				        strings.Contains(history.QueryText, query)
```

File: pkg/resources/function.go (hunk 2 above)
Comment (sfc-gh-jmichalak) [reply pkg/resources/function.go:182]: It's not, but it's passthrough, so I'm resolving.

File: pkg/acceptance/helpers/information_schema_client.go (hunk 7 above)
Comment (sfc-gh-jmichalak) [reply pkg/acceptance/helpers/information_schema_client.go]: Looks good now, I meant that we can use LIMIT 1 in GetQueryHistoryByQueryId, right? But that's a nit.

**Review by sfc-gh-jmichalak — APPROVED** [review]

**Review by sfc-gh-asawicki — DISMISSED** [review]

**Review by sfc-gh-jmichalak — APPROVED** [review]

**Pull request merged**